syntax = "proto3";

// A Unary streaming RPC.
package unary;

message InventoryRecord {
  string Inventory_ID = 1;
  string Name = 2;
  string Description = 3;
  double Price = 4;
  int32 Quantity_in_Stock = 5;
  int32 Inventory_Value = 6;
  int32 Reorder_Level = 7;
  int32 Quantity_in_Reorder = 8;
  bool Discontinued = 9;
}

message InventoryRequest {
  string id = 1;
}

message InventorySearchRequest {
  string key_name = 1;
  string key_value = 2;
}

message InventoryRangeRequest {
  string key_name = 1;
  string key_value_start = 2;
  string key_value_end = 3;
}

message DistributionRequest {
  string key_name = 1;
  double percentile = 2;
}

message UpdateRequest {
  string key_name = 1;
  string key_value = 2;
  string val_name = 3;
  string val_val_new = 4;
}

message DistributionResponse {
  double value = 1;
}

message UpdateResponse {
  bool success = 1;
}

service InventoryService {
  rpc searchByID (InventoryRequest) returns (InventoryRecord);
  rpc search (InventorySearchRequest) returns (InventoryRecord);
  rpc searchRange (InventoryRangeRequest) returns (stream InventoryRecord);
  rpc getDistribution (DistributionRequest) returns (DistributionResponse);
  rpc update (UpdateRequest) returns (UpdateResponse);
}
